### PR TITLE
Flutter External Texture Based On OpenGL Texture Underlying Share Context

### PR DESCRIPTION
**Pull request by kaisa695275735**
This  is a new PR about External Texture,  the old one(https://github.com/flutter/engine/pull/11276) is closed because of my mistaken git operation.

### CODE REVIEW

**Comment by kaisa695275735** [comment]
Hi  @dnfield @gaaclarke   We've added the unit test code in testing/scenarios_app for IOS, which is similar to the platform view example.
For Android ,we've  finished the demo code to be tested, but did not find any good way for testing benchmark

**Comment by dnfield** [comment]
We don't have as much wired up for Android on CI just yet.  Before we look at this, can you work on the merge conflicts?  @gw280 just landed a patch that is causing some conflicts with this one.

**Comment by kaisa695275735** [comment]
> We don't have as much wired up for Android on CI just yet. Before we look at this, can you work on the merge conflicts? @gw280 just landed a patch that is causing some conflicts with this one.

Conflicts are resolved already.

**Comment by gw280** [comment]
> > We don't have as much wired up for Android on CI just yet. Before we look at this, can you work on the merge conflicts? @gw280 just landed a patch that is causing some conflicts with this one.
> 
> Conflicts are resolved already.

My change got backed out because it was failing LUCI, so I'm currently working on reworking that. When are you hoping to merge this change?

File: shell/platform/android/platform_view_android_jni.cc [hunk 1]
```diff
@@ -444,6 +446,52 @@ static void RegisterTexture(JNIEnv* env,
...
+                                 jlong shell_holder,
+                                 jlong texture_id,
+                                 jlong share_texture_id) {
+  ANDROID_SHELL_HOLDER->GetPlatformView()->RegisterExternalShareTexture(
+      static_cast<int64_t>(texture_id),       //
+      static_cast<int64_t>(share_texture_id)  //
+  );
+}
+
+static jobject GetShareContext(JNIEnv* env,
```
Comment (gaaclarke) [inline shell/platform/android/platform_view_android_jni.cc]: Should this method have a try block to catch c++ exceptions and translate them to java exceptions?  There might be helpers for that.  I forget if we allow exceptions in our c++ code.

File: shell/platform/darwin/ios/ios_external_texture_gl_share_context.h [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+#ifndef FLUTTER_SHELL_PLATFORM_IOS_EXTERNAL_TEXTURE_SHARE_CONTEXT_H_
+#define FLUTTER_SHELL_PLATFORM_IOS_EXTERNAL_TEXTURE_SHARE_CONTEXT_H_
+
+#include "flutter/flow/texture.h"
+#include "flutter/fml/platform/darwin/cf_utils.h"
+#include "flutter/shell/platform/darwin/ios/framework/Headers/FlutterTexture.h"
+
+namespace flutter {
+// This is a new way of external texture.
```
Comment (gaaclarke) [inline shell/platform/darwin/ios/ios_external_texture_gl_share_context.h]: Seems like there might be something missing after the word 'of' here.

File: shell/platform/darwin/ios/ios_external_texture_gl_share_context.h [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+// Unlike the original external texture which copy pixel buffer from an object that impelement
```
Comment (gaaclarke) [inline shell/platform/darwin/ios/ios_external_texture_gl_share_context.h]: *which copies the pixel buffer
*that implements the FlutterTexture protocol
*and creates a texture reference
*pixel buffer, this solution will
impelementing -> implementing

**Comment by kaisa695275735** [comment]
> > > We don't have as much wired up for Android on CI just yet. Before we look at this, can you work on the merge conflicts? @gw280 just landed a patch that is causing some conflicts with this one.
> > 
> > 
> > Conflicts are resolved already.
> 
> My change got backed out because it was failing LUCI, so I'm currently working on reworking that. When are you hoping to merge this change?

We may not merge this change until the code reviews are all resolved.

File: shell/platform/android/platform_view_android_jni.cc (hunk 1 above)
Comment (kaisa695275735) [reply shell/platform/android/platform_view_android_jni.cc]: it seems that exceptions was disabled. 
We got a compile error when we add a try block:
error: cannot use 'try' with exceptions disabled

File: shell/platform/android/android_external_texture_gl_share_context.cc [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+                                               const SkRect& bounds,
+                                               bool freeze,
+                                               GrContext* context) {
+  GrGLTextureInfo textureInfo = {GL_TEXTURE_EXTERNAL_OES, texture_id_,
+                                 GL_RGBA8_OES};
+
+  textureInfo.fTarget = GL_TEXTURE_2D;
+  transform.setIdentity();
+
+  GrBackendTexture backendTexture(1, 1, GrMipMapped::kNo, textureInfo);
```
[the remark this answers is not in the document]
Comment (kaisa695275735) [reply shell/platform/android/android_external_texture_gl_share_context.cc]: The variables "1" here is a normalized value of the width and height of bounds for the canvas's transform operation.
The transform matrix is updated from the SurfaceTexture which has transform information.
Here we use OpenGL Texture without transform information, so we remove this code 
and  change the backendTexture variables to bounds.width() and bounds.height().

File: shell/platform/android/android_external_texture_gl_share_context.cc [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  sk_sp<SkImage> image = SkImage::MakeFromTexture(
+      canvas.getGrContext(), backendTexture, kTopLeft_GrSurfaceOrigin,
+      kRGBA_8888_SkColorType, kPremul_SkAlphaType, nullptr);
+  if (image) {
```
[the remark this answers is not in the document]
Comment (kaisa695275735) [reply shell/platform/android/android_external_texture_gl_share_context.cc]: YES  we  add a  check for image to be null, just as it is on IOS  side.

File: shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java [hunk 6]
```diff
@@ -110,6 +111,24 @@ public SurfaceTextureEntry createSurfaceTexture() {
     return entry;
   }
 
+  @Override
+  public TextureRegistry.ShareTextureEntry createShareTexture(long shareTextureID) {
+    final ShareTextureRegistryEntry entry = new ShareTextureRegistryEntry(nextTextureId.getAndIncrement(),shareTextureID);
```
[the remark this answers is not in the document]
Comment (kaisa695275735) [reply shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java]: Fix it!

**Comment by kangwang1988** [comment]
@dnfield 
@kaisa695275735  is my colleague focusing on video related works.
Should we wait for more members to review it or it's about ready to be merged?

**Comment by dnfield** [comment]
Thanks!  It needs to have a LGTM before meeting.  I can try to take a look today as well but it looks like @gaaclarke is on it.

**Comment by kaisa695275735** [comment]
@gaaclarke we have make our changes. 
please take a look! Thank you very much!

File: shell/platform/darwin/ios/platform_view_ios.h [hunk 7]
```diff
@@ -37,13 +37,17 @@ class PlatformViewIOS final : public PlatformView {
...
+  void RegisterExternalShareTexture(int64_t id, NSObject<FlutterShareTexture>* texture);
+
   fml::scoped_nsprotocol<FlutterTextInputPlugin*> GetTextInputPlugin() const;
 
   void SetTextInputPlugin(fml::scoped_nsprotocol<FlutterTextInputPlugin*> plugin);
 
   // |PlatformView|
   void SetSemanticsEnabled(bool enabled) override;
 
+  EAGLSharegroup* GetGLShareGroup();
```
Comment (gaaclarke) [inline shell/platform/darwin/ios/platform_view_ios.h]: Leaking OpenGL in renderer agnostic code.

File: shell/platform/darwin/ios/ios_surface.h [hunk 8]
```diff
@@ -39,6 +39,8 @@ class IOSSurface {
   virtual std::unique_ptr<Surface> CreateGPUSurface(
       GrContext* gr_context = nullptr) = 0;
 
+  virtual EAGLSharegroup* GetGLShareGroup() = 0;
```
Comment (gaaclarke) [inline shell/platform/darwin/ios/ios_surface.h]: Leaking OpenGL into renderer agnostic code.

File: shell/platform/darwin/ios/framework/Headers/FlutterTexture.h [hunk 9]
```diff
@@ -8,20 +8,32 @@
...
 - (CVPixelBufferRef _Nullable)copyPixelBuffer;
 @end
 
 FLUTTER_EXPORT
 @protocol FlutterTextureRegistry <NSObject>
 - (int64_t)registerTexture:(NSObject<FlutterTexture>*)texture;
+- (int64_t)registerShareTexture:(NSObject<FlutterShareTexture>*)texture;
 - (void)textureFrameAvailable:(int64_t)textureId;
 - (void)unregisterTexture:(int64_t)textureId;
+- (EAGLSharegroup*)getShareGroup;
```
Comment (gaaclarke) [inline shell/platform/darwin/ios/framework/Headers/FlutterTexture.h]: Leaking OpenGL into renderer agnostic code.

File: shell/platform/darwin/ios/ios_external_texture_gl_share_context.h [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+#include "flutter/flow/texture.h"
+#include "flutter/fml/platform/darwin/cf_utils.h"
+#include "flutter/shell/platform/darwin/ios/framework/Headers/FlutterTexture.h"
+
+namespace flutter {
+// This is another solution for Flutter's external texture.
+// Unlike the original external texture which copies pixel buffer from an object that implements
+// FlutterTexture protocol, and creates a texture reference using the pixel buffer, this solution
+// will copy the OpenGL texture directly from the object implementing FlutterShareTexture protocol.
+class IOSExternalTextureShareContext : public flutter::Texture {
```
Comment (gaaclarke) [inline shell/platform/darwin/ios/ios_external_texture_gl_share_context.h]: This should probably be named IOSExternalShareContextTexture.  The type is usually placed at the end of the name.

**Comment by gaaclarke** [comment]
I drew up the iOS portion of the PR:
share-context

**Comment by cbracken** [comment]
@kaisa695275735 just checking in to see if you're still working on this PR. If so, no worries!

**Comment by zhangruiyu** [comment]
????

**Comment by cbracken** [comment]
@kaisa695275735 is PR still work-in-progress? Looks like we're waiting on fixes to address @gaaclarke 's comments.

**Comment by kaisa695275735** [comment]
> @kaisa695275735 is PR still work-in-progress? Looks like we're waiting on fixes to address @gaaclarke 's comments.

sorry for my later reply. I took a long vocation last month.
we are trying to use CVOpenGLESTextureCache to implement a filter camera demo, and we find this scheme does have the same CPU level with our share texture scheme.
So all the code for iOS is not needed.
We are trying to use SurfaceTexture scheme for Android, if they have the same CPU level,
this PR  may be closed.

**Pull request closed without merging**